### PR TITLE
[MIR] Implement extern call support

### PR DESCRIPTION
**Pull request by nagisa**
Supersedes https://github.com/rust-lang/rust/pull/30517
Fixes https://github.com/rust-lang/rust/issues/29575

cc @luqmana 
r? @nikomatsakis

### CODE REVIEW

**Comment by luqmana** [comment]
LGTM. :+1:

**Comment by Manishearth** [comment]
This or #30848 caused http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2673/steps/test/logs/stdio

File: src/test/run-pass/mir_trans_calls.rs [hunk 1]
```diff
@@ -88,6 +89,31 @@ fn test8() -> isize {
...
+fn test9() -> u32 {
+    simple_extern(41, (42, 43))
+}
+
+extern {
+    fn sprintf(_: *mut raw::c_char, _: *const raw::c_char, ...) -> raw::c_int;
+}
+
+#[rustc_mir]
+fn test10(i: i32, j: i32, k: i32) -> Vec<raw::c_char> {
```
Comment (nagisa) [inline src/test/run-pass/mir_trans_calls.rs]: So… this test seems to cause STATUS_HEAP_CORRUPTION… Hmm.

**Comment by nagisa** [comment]
@bors r-

The test causes some sort of heap corruption on Windows.

**Comment by retep998** [comment]
Testing that test in the x86_64 msvc nightly without the`#[rustc_mir]`, I failed to replicate any sort of heap corruption.

**Comment by nagisa** [comment]
@bors try- 

(we have no MSVC try bots)

**Comment by nagisa** [comment]
@bors r=nikomatsakis

(adjusted the test with MSVC workarounds, sending to buildbot to see how it goes this time around)

**Comment by nagisa** [comment]
@bors r- try- r=nmatsakis retry force p=1 38bed95

Hmm.

**Comment by nagisa** [comment]
The reason has been narrowed down. The original trans also fails to translate that test, as has been shown by https://github.com/rust-lang/rust/issues/30921 and [its result](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2692).

**Comment by nagisa** [comment]
Okay, so the reason ends up being [LLVM itself](https://llvm.org/bugs/show_bug.cgi?id=26159). I’ll just ignore the test in question on msvc like it is done in the other related test for now.

**Comment by nagisa** [comment]
I couldn’t reproduce the original failure with VC2013-based rustc locally – the test runs, produces expected results and quits successfully. In the current iteration, the test is disabled on msvc (just like the regular-trans variadic-ffi) and the PR should pass through build bots this time. That being said, the original error is still unsettling and I would like to try sending re-enabled test over to msvc bots again. Is that fine?

(Trivia: I once had encountered a SIGSEGV on a unix builder, which turned out to be spurious, before, so the original failure could also be spurious in some way? /me shrugs)

**Comment by nikomatsakis** [comment]
@nagisa 

> In the current iteration, the test is disabled on msvc (just like the regular-trans variadic-ffi) 

OK

> That being said, the original error is still unsettling and I would like to try sending re-enabled test over to msvc bots again

OK -- do you want to land this PR, then open another that re-enables the test? We can also use the try bots, in that case.

**Comment by nagisa** [comment]
> OK -- do you want to land this PR, then open another that re-enables the test? We can also use the try bots, in that case.

I think what I’d like to do is try to land this with the test enabled now, and then, provided this PR manages to land, disable it for MSVC in a follow-up PR. The test has to stay disabled, because otherwise it will break everybody’s MSVC-2015-based setups (remember, in that particular version of MSVC sprintf does not exist as a function).

Does that sound good to you?

**Comment by nagisa** [comment]
Upon suggestion of @dotdash I changed the test to use our own variadic function (defined inside rt/rust_test_helpers.c) instead so we don’t have to worry about the MSVC strangeness.

re-r? @nikomatsakis

**Pull request merged**